Building Jupyterlab plugin from source does not pull working directory changes

Previously, when I built the `perspective-jupyterlab` plugin, it would pull in changes from other packages in Perspective—thus I could make a change in `perspective-viewer` and see it reflected in Jupyterlab, etc.

Since #1231, building the Jupyterlab plugin no longer pulls in changes from the local working directory, an issue which I fixed [here](https://github.com/finos/perspective/pull/1289/files#diff-13040e3ab93f2f6f662c5a9cbe13cdf094e09f4c21f9123c9c0acdef589cd9c4R25) as part of #1289, before the Jlab 3 PR (#1294) was merged. The issue with it not resolving working directory changes was due to `externals: [/^[a-z0-9@]/],` which meant that `perspective` was not being bundled into the Jupyterlab plugin, and so the plugin could only pull the disted version of Perspective (assuming my understanding of Webpack in this case is correct).

I tried to implement the same fix on top of #1294, but `jupyter lab build` fails with this error:

```javascript
[webpack-cli] ModuleNotFoundError: Module not found: Error: Can't resolve 'fs' in '/Library/Frameworks/Python.framework/Versions/3.7/share/jupyter/lab/staging/node_modules/@finos
/perspective-jupyterlab/dist'
    at /Library/Frameworks/Python.framework/Versions/3.7/share/jupyter/lab/staging/node_modules/webpack/lib/Compilation.js:1570:28
    at /Library/Frameworks/Python.framework/Versions/3.7/share/jupyter/lab/staging/node_modules/webpack/lib/NormalModuleFactory.js:648:13
    at eval (eval at create (/Library/Frameworks/Python.framework/Versions/3.7/share/jupyter/lab/staging/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:8:1)
    at /Library/Frameworks/Python.framework/Versions/3.7/share/jupyter/lab/staging/node_modules/webpack/lib/NormalModuleFactory.js:233:22
    at eval (eval at create (/Library/Frameworks/Python.framework/Versions/3.7/share/jupyter/lab/staging/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:7:1)
    at /Library/Frameworks/Python.framework/Versions/3.7/share/jupyter/lab/staging/node_modules/webpack/lib/NormalModuleFactory.js:357:22
    at /Library/Frameworks/Python.framework/Versions/3.7/share/jupyter/lab/staging/node_modules/webpack/lib/NormalModuleFactory.js:116:11
    at /Library/Frameworks/Python.framework/Versions/3.7/share/jupyter/lab/staging/node_modules/webpack/lib/NormalModuleFactory.js:577:24
    at /Library/Frameworks/Python.framework/Versions/3.7/share/jupyter/lab/staging/node_modules/webpack/lib/NormalModuleFactory.js:721:5
    at finishWithoutResolve (/Library/Frameworks/Python.framework/Versions/3.7/share/jupyter/lab/staging/node_modules/enhanced-resolve/lib/Resolver.js:296:11)
resolve 'fs' in '/Library/Frameworks/Python.framework/Versions/3.7/share/jupyter/lab/staging/node_modules/@finos/perspective-jupyterlab/dist'
  Parsed request is a module
  using description file: /Library/Frameworks/Python.framework/Versions/3.7/share/jupyter/lab/staging/node_modules/@finos/perspective-jupyterlab/package.json (relative path: ./dist)
    Field 'browser' doesn't contain a valid alias configuration
    resolve as module
      /Library/Frameworks/Python.framework/Versions/3.7/share/jupyter/lab/staging/node_modules/@finos/perspective-jupyterlab/dist/node_modules doesn't exist or is not a directory
      /Library/Frameworks/Python.framework/Versions/3.7/share/jupyter/lab/staging/node_modules/@finos/perspective-jupyterlab/node_modules doesn't exist or is not a directory
      /Library/Frameworks/Python.framework/Versions/3.7/share/jupyter/lab/staging/node_modules/@finos/node_modules doesn't exist or is not a directory
      /Library/Frameworks/Python.framework/Versions/3.7/share/jupyter/lab/staging/node_modules/node_modules doesn't exist or is not a directory
      looking for modules in /Library/Frameworks/Python.framework/Versions/3.7/share/jupyter/lab/staging/node_modules
        single file module
          using description file: /Library/Frameworks/Python.framework/Versions/3.7/share/jupyter/lab/staging/package.json (relative path: ./node_modules/fs)
            no extension
              Field 'browser' doesn't contain a valid alias configuration
              /Library/Frameworks/Python.framework/Versions/3.7/share/jupyter/lab/staging/node_modules/fs doesn't exist
            .js
              Field 'browser' doesn't contain a valid alias configuration
              /Library/Frameworks/Python.framework/Versions/3.7/share/jupyter/lab/staging/node_modules/fs.js doesn't exist
            .json
              Field 'browser' doesn't contain a valid alias configuration
              /Library/Frameworks/Python.framework/Versions/3.7/share/jupyter/lab/staging/node_modules/fs.json doesn't exist
            .wasm
              Field 'browser' doesn't contain a valid alias configuration
              /Library/Frameworks/Python.framework/Versions/3.7/share/jupyter/lab/staging/node_modules/fs.wasm doesn't exist
        /Library/Frameworks/Python.framework/Versions/3.7/share/jupyter/lab/staging/node_modules/fs doesn't exist
      /Library/Frameworks/Python.framework/Versions/3.7/share/jupyter/lab/node_modules doesn't exist or is not a directory
      /Library/Frameworks/Python.framework/Versions/3.7/share/jupyter/node_modules doesn't exist or is not a directory
      /Library/Frameworks/Python.framework/Versions/3.7/share/node_modules doesn't exist or is not a directory
      /Library/Frameworks/Python.framework/Versions/3.7/node_modules doesn't exist or is not a directory
      /Library/Frameworks/Python.framework/Versions/node_modules doesn't exist or is not a directory
      /Library/Frameworks/Python.framework/node_modules doesn't exist or is not a directory
      /Library/Frameworks/node_modules doesn't exist or is not a directory
      /Library/node_modules doesn't exist or is not a directory
      /node_modules doesn't exist or is not a directory
error Command failed with exit code 2.
```

I have added `node: {fs: "empty"}` to `plugin.config.js` and the same issue happens, and we are definitely not importing 'fs' anywhere in our disted assets - @timkpaine not sure if you have insight into this issue/have a good direction to look.

Yeah I'm on jupyterlab 3.0.5 doing `jupyter labextension install @finos/perspective-jupyterlab`. And I get 
`[webpack-cli] ModuleNotFoundError: Module not found: Error: Can't resolve 'fs' in '.../anaconda3/envs/test/share/jupyter/lab/staging/node_modules/@finos/perspective-jupyterlab/dist'`

--- a/docs/md/development.md
+++ b/docs/md/development.md
@@ -91,6 +91,34 @@ To install this specific version of Emscripten:
 ./emsdk install 2.0.6
 ```
 
+### `perspective-jupyterlab`
+
+To install the Jupyterlab plugin from your local working directory, give
+`jupyter labextension install` the path to the `perspective-jupyterlab`
+package:
+
+```bash
+jupyter labextension install ./packages/perspective-jupyterlab
+```
+
+Afterwards, you should see it listed as a "local extension" when you run
+`jupyter labextension list`.
+
+Because we do not inline Perspective into the Jupyterlab plugin, your local
+changes will not show up in Jupyterlab **unless** you use `yarn link`
+according to the directions below:
+
+1. Ensure that your Jupyterlab is built by running `jupyter lab build`.
+2. Inside each directory in `packages`, run [`yarn link`](https://classic.yarnpkg.com/en/docs/cli/link).
+This will create a symlink to your local build that we will use inside Jupyterlab.
+3. From the Perspective root, run `yarn jlab_link`. This is a script that will
+find your Jupyterlab installation and tell Jupyterlab to use these symlinks
+when it looks for Perspective packages, instead of fetching them from NPM.
+4. When you make a local change, make sure you run `yarn build` **and**
+`jupyter lab build` so that it fetches the newest changes.
+5. Whenever you run `jupyter lab clean`, you will need to run `yarn jlab_link`
+again to re-register the symlinks.
+
 ## `perspective-python`
 
 To build the Python library, first configure your project to Python via

--- a/docs/md/python.md
+++ b/docs/md/python.md
@@ -68,8 +68,10 @@ pip install perspective-python
 `PerspectiveWidget` is a JupyterLab widget that implements the same API as
 `<perspective-viewer>`, allowing for fast, intuitive
 transformations/visualizations of various data formats within JupyterLab.
-To use it, make sure you have installed `perspective-python`
-and then install the extension from the Jupyter lab extension directory
+
+`PerspectiveWidget` is compatible with Jupyterlab 3. To use it, make sure you
+have installed `perspective-python` and then install the extension from the
+Jupyter lab extension directory:
 
 ```bash
 jupyter labextension install @finos/perspective-jupyterlab

--- a/docs/static/css/custom.css
+++ b/docs/static/css/custom.css
@@ -203,7 +203,6 @@ ol li {
 ol li::before {
     content: counter(list-counter) ".";
     font-weight: bold;
-    font-size: 2rem;
     margin-right: 0.5rem;
     font-family: "Public Sans", Arial, Helvetica, sans-serif;
     line-height: 1;

--- a/docs/static/css/material.dark.css
+++ b/docs/static/css/material.dark.css
@@ -167,6 +167,7 @@ perspective-viewer,
   --color: #cfd8dc;
   --plugin--background: #2f3136;
   --select--background: url(data:image/svg+xml;base64,PD94bWwgdmVyc2lvbj0iMS4wIiBlbmNvZGluZz0idXRmLTgiPz4KPCEtLSBHZW5lcmF0b3I6IEFkb2JlIElsbHVzdHJhdG9yIDIwLjAuMCwgU1ZHIEV4cG9ydCBQbHVnLUluIC4gU1ZHIFZlcnNpb246IDYuMDAgQnVpbGQgMCkgIC0tPgo8c3ZnIHZlcnNpb249IjEuMSIgaWQ9IkxheWVyXzEiIHhtbG5zPSJodHRwOi8vd3d3LnczLm9yZy8yMDAwL3N2ZyIgeG1sbnM6eGxpbms9Imh0dHA6Ly93d3cudzMub3JnLzE5OTkveGxpbmsiIHg9IjBweCIgeT0iMHB4IgoJIHZpZXdCb3g9IjAgMCA0LjkgMTAiIHN0eWxlPSJlbmFibGUtYmFja2dyb3VuZDpuZXcgMCAwIDQuOSAxMDsiIHhtbDpzcGFjZT0icHJlc2VydmUiPgo8c3R5bGUgdHlwZT0idGV4dC9jc3MiPgoJLnN0MHtmaWxsOiM0NDQ0NDQ7fQo8L3N0eWxlPgo8dGl0bGU+YXJyb3dzPC90aXRsZT4KPHBvbHlnb24gY2xhc3M9InN0MCIgcG9pbnRzPSIxLjQsNC43IDIuNSwzLjIgMy41LDQuNyAiLz4KPHBvbHlnb24gY2xhc3M9InN0MCIgcG9pbnRzPSIzLjUsNS4zIDIuNSw2LjggMS40LDUuMyAiLz4KPC9zdmc+Cg==) no-repeat 95% 50%;
+  --select--background-color: #242526;
   --inactive--color: #666;
   --active--background: #135d78;
   --active--color: #22a0ce;

--- a/docs/static/js/animation.js
+++ b/docs/static/js/animation.js
@@ -42,16 +42,16 @@ var freq = 1,
     freqdir = 1,
     elem;
 
-function update() {
+function update(table) {
     var viewport_height = document.documentElement.clientHeight;
     if (viewport_height - window.scrollY > 0) {
-        elem.update([newRow(), newRow(), newRow()]);
+        table.update([newRow(), newRow(), newRow()]);
     }
     if (freq === 0) {
-        setTimeout(update, 3000);
+        setTimeout(() => update(table), 3000);
         freqdir = 1;
     } else {
-        setTimeout(update, Math.max(20, 200 / freq));
+        setTimeout(() => update(table), Math.max(20, 200 / freq));
     }
     if (freq > 60) {
         freqdir = -1;
@@ -138,13 +138,13 @@ window.addEventListener("WebComponentsReady", function() {
         data.push(newRow());
     }
     elem = Array.prototype.slice.call(document.querySelectorAll("perspective-viewer"))[0];
-    var worker = elem.worker;
+    var worker = perspective.worker();
     var tbl = worker.table(data, {index: "id"});
     elem.load(tbl);
     elem._toggle_config();
 
     setTimeout(function() {
-        update(0);
+        update(tbl, 0);
     });
 
     document.querySelector("#grid").addEventListener("mouseenter", () => select("#grid"));
@@ -175,7 +175,10 @@ window.addEventListener("WebComponentsReady", function() {
             plugin: "d3_heatmap",
             "column-pivots": ["Sub-Category"],
             "row-pivots": ["State"],
-            sort: [["Sales","desc"],["Profit","col desc"]],
+            sort: [
+                ["Sales", "desc"],
+                ["Profit", "col desc"]
+            ],
             columns: ["Profit"],
             aggregates: {Profit: "low"}
         });

--- a/package.json
+++ b/package.json
@@ -133,6 +133,7 @@
         "fix:python": "node scripts/lint_python.js --fix",
         "fix": "npm-run-all --silent fix:*",
         "toggle_puppeteer": "node scripts/toggle_puppeteer.js",
-        "version": "python3 python/perspective/scripts/write_version.py && git add python/perspective/perspective/core/_version.py"
+        "version": "python3 python/perspective/scripts/write_version.py && git add python/perspective/perspective/core/_version.py",
+        "jlab_link": "node scripts/jlab_link.js"
     }
 }

--- a/packages/perspective-jupyterlab/package.json
+++ b/packages/perspective-jupyterlab/package.json
@@ -32,6 +32,7 @@
         "version": "yarn build"
     },
     "dependencies": {
+        "@finos/perspective": "^0.6.0",
         "@finos/perspective-viewer": "^0.6.0",
         "@finos/perspective-viewer-d3fc": "^0.6.0",
         "@finos/perspective-viewer-datagrid": "^0.6.0",

--- a/packages/perspective/src/js/perspective.js
+++ b/packages/perspective/src/js/perspective.js
@@ -1799,7 +1799,6 @@ export default function(Module) {
                 throw new Error("WebAssembly not supported");
             } else {
                 console.log("Loading wasm");
-                console.log(msg.buffer.byteLength);
                 __MODULE__({
                     wasmBinary: msg.buffer,
                     wasmJSMethod: "native-wasm"

--- a/scripts/jlab_link.js
+++ b/scripts/jlab_link.js
@@ -1,0 +1,71 @@
+/******************************************************************************
+ *
+ * Copyright (c) 2017, the Perspective Authors.
+ *
+ * This file is part of the Perspective library, distributed under the terms of
+ * the Apache License 2.0.  The full license can be found in the LICENSE file.
+ *
+ */
+const path = require("path");
+const {existsSync} = require("fs");
+const {execute, execute_return} = require("./script_utils.js");
+
+// Packages listed as dependencies & dev dependencies inside Jupyterlab plugin
+const packages = ["perspective", "perspective-viewer", "perspective-viewer-datagrid", "perspective-viewer-d3fc", "perspective-test", "perspective-webpack-plugin"].map(pkg => `@finos/${pkg}`);
+
+/**
+ * In order for Jupyterlab to pick up changes to @finos/perspective-* in the
+ * local working directory, we use `yarn link` to force Jupyterlab to look
+ * locally for all `perspective` packages, instead of pulling them from NPM
+ * during `jupyter lab install`. This script should be run whenever
+ * `jupyter lab clean` has been run, as cleaning the Jupyterlab dir will
+ * also break links.
+ *
+ * This allows us to distribute just the Jupyterlab plugin without inlining
+ * the rest of Perspective, but also allows for developers to see their local
+ * changes picked up in their current Jupyterlab installation.
+ *
+ * To set up your Jupyterlab environment with the local
+ * `perspective-jupyterlab` plugin, run
+ * `jupyter labextension install ./packages/perspective-jupyterlab` to link
+ * to the local plugin.
+ */
+(async function() {
+    try {
+        // check if @finos packages are already linked
+        const link_path = path.resolve(process.env.HOME, ".config/yarn/link", "@finos");
+
+        if (!existsSync(link_path)) {
+            throw Error(`No @finos links found in ${link_path} - run \`yarn link\` inside all Perspective packages first.`);
+        }
+
+        const paths = await execute_return("jupyter lab path");
+        const app_path = paths["stdout"]
+            .split("\n")[0]
+            .split(":")[1]
+            .trim();
+        const staging_path = path.resolve(app_path, "staging");
+
+        if (!existsSync(staging_path)) {
+            throw Error("Jupyterlab staging path does not exist - run `jupyter lab build` first.");
+        }
+
+        process.chdir(staging_path);
+
+        // Run yarn link inside jlab staging
+        for (const pkg of packages) {
+            execute(`yarn link ${pkg}`);
+        }
+
+        // Build from Perspective root
+        process.chdir(path.resolve(__dirname, ".."));
+        console.log("--------------------------");
+        execute("jupyter lab build");
+
+        console.log("--------------------------");
+        console.log("Jupyterlab should now have all changes from the current working directory. To pick up new changes, run `yarn build` and `jupyter lab build`.");
+    } catch (e) {
+        console.error(e);
+        process.exit(1);
+    }
+})();

--- a/scripts/script_utils.js
+++ b/scripts/script_utils.js
@@ -13,6 +13,8 @@ const execSync = require("child_process").execSync;
 const _path = require("path");
 const fs = require("fs");
 const rimraf = require("rimraf");
+const {promisify} = require("util");
+
 const isWin = process.platform === "win32";
 
 /*******************************************************************************
@@ -69,6 +71,16 @@ const execute_throw = cmd => {
     }
     execSync(cmd, {stdio: "inherit"});
 };
+
+const execute_return = async cmd => {
+    if (process.argv.indexOf("--debug") > -1) {
+        console.log(`$ ${cmd}`);
+    }
+
+    const ex = promisify(require("child_process").exec);
+
+    return await ex(cmd);
+}
 
 /*******************************************************************************
  *
@@ -185,6 +197,18 @@ exports.execute = (strings, ...args) => execute(Array.isArray(strings) ? bash(st
  * execute`run -t${1} -u"${undefined}" task`;
  */
 exports.execute_throw = (strings, ...args) => execute_throw(Array.isArray(strings) ? bash(strings, ...args) : strings);
+
+/**
+ * Just like `execute`, except it will return the output of the command
+ * it runs.
+ *
+ * @async
+ * @param {string} expression a bash command to be templated.
+ * @returns {string} A command with the missing argument's flags removed.
+ * @example
+ * execute`run -t${1} -u"${undefined}" task`;
+ */
+exports.execute_return = async (strings, ...args) => await execute_return(Array.isArray(strings) ? bash(strings, ...args) : strings);
 
 /**
  * Returns the value after this command-line flag, or `true` if it is the last

